services/src/main/java/org/keycloak/services/resources/admin/ClientsResource.java

```diff
@@ -91,15 +89,22 @@ public ClientsResource(RealmModel realm, AdminPermissionEvaluator auth, AdminEve
      *
      * @param clientId filter by clientId
      * @param viewableOnly filter clients that cannot be viewed in full by admin
+     * @param search whether this is a search query or a getClientById query
+     * @param firstResult the first result
+     * @param maxResults the max results to return
      */
     @GET
     @Produces(MediaType.APPLICATION_JSON)
     @NoCache
-    public List<ClientRepresentation> getClients(@QueryParam("clientId") String clientId, @QueryParam("viewableOnly") @DefaultValue("false") boolean viewableOnly) {
+    public List<ClientRepresentation> getClients(@QueryParam("clientId") String clientId,
+                                                 @QueryParam("viewableOnly") @DefaultValue("false") boolean viewableOnly,
+                                                 @QueryParam("search") @DefaultValue("false") boolean search,
```

Comment:
I think you may add a new method in `public interface ClientsResource` with all theses params. I'm pretty sure about that but wait for @stianst  comfirmation. Also for me you should rewrite this function to separate the `findByClientId` and the `findAll` implementation and so your `search` query param become a `String` empty by default to be coherent with other search function in the application ( you can look on the search in `GroupsResource` ), on this wait for @stianst advice too because your solution works but it's not coherent with others endpoints.

Reply:
I understand the need, it would be nice to have this separation. However, I think that this sort of improvement could come as a separate refactoring task / story, it's not particularly in the scope of the initial development.

I'd be happy to contribute more, once we get this one rolling, as it's been a couple of months since it's been open already.

Reply:
For the second point about separate the function it's not just refactoring, if you define your api with `search` param as boolean and that users start to use it like that, we will have to keep it like that or release an other version different of the api. It sound a bit weird to me. But regarding the current code I don't see an easy solution, this getClientById should probably not have been implemented like that at the beginining because the current params `clientId` is for an `exact` search which can just return one result, should have be a path param probably in a different endpoint !

So I don't know maybe @stianst will have an idea I don't know what happend if we have two implementation using `viewableOnly` on the same path, we will probably have a jaxrs conflict somewhere.

Anyway if you keep things like that, in the `ClientsResource` interface you should add 

```
    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public List<ClientRepresentation> searchByClientId(@QueryParam("clientId") String clientId,
                                                 @QueryParam("viewableOnly") @DefaultValue("false") boolean viewableOnly,
                                                 @QueryParam("search") @DefaultValue("false") boolean search,
                                                 @QueryParam("first") Integer firstResult,
                                                 @QueryParam("max") Integer maxResults) ;
```

Reply:
The interface ClientsResource in admin-client should have a new method, with the extra param.

For refactoring of the Admin REST APIs - the Admin REST APIs are not extremely nice as they are and there are lots of inconsistencies. What we plan to do when we get time to do it is to introduce new smaller versioned APIs. So there will be a clients API, a users API, etc. each with their own versions.